xwiki-platform-core/xwiki-platform-store/xwiki-platform-store-filesystem-oldcore/src/main/java/org/xwiki/store/filesystem/internal/DefaultTemporaryAttachmentManager.java

```diff
@@ -0,0 +1,261 @@
+
+    protected static final long UPLOAD_DEFAULT_MAXSIZE = 33554432L;
+
+    private Map<String, TemporaryAttachmentSession> temporaryAttachmentSessionMap;
+
+    @Inject
+    private CacheManager cacheManager;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private EntityReferenceSerializer<String> stringEntityReferenceSerializer;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        this.temporaryAttachmentSessionMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void dispose() throws ComponentLifecycleException
+    {
+        this.temporaryAttachmentSessionMap.values().forEach(TemporaryAttachmentSession::dispose);
+    }
+
+    @Override
+    public void sessionCreated(HttpSessionEvent httpSessionEvent)
+    {
+        // Nothing should be done here.
+    }
+
+    /**
+     * Accessor to the map of temporary attachment sessions. Mainly for test purpose.
+     *
+     * @return the map of temporary attachment session.
+     */
+    protected Map<String, TemporaryAttachmentSession> getTemporaryAttachmentSessionMap()
+    {
+        return temporaryAttachmentSessionMap;
+    }
+
+    private String getSessionId()
+    {
+        XWikiContext context = this.contextProvider.get();
+        return context.getRequest().getSession().getId();
+    }
+
+    @Override
+    public void sessionDestroyed(HttpSessionEvent httpSessionEvent)
+    {
+        String sessionId = httpSessionEvent.getSession().getId();
+        if (this.temporaryAttachmentSessionMap.containsKey(sessionId)) {
+            this.temporaryAttachmentSessionMap.get(sessionId).dispose();
+            this.temporaryAttachmentSessionMap.remove(sessionId);
+        }
+    }
+
+    private String getCacheName(String sessionId, DocumentReference documentReference)
+    {
+        return String.format("temp.attachment.%s_%s", sessionId,
+            this.stringEntityReferenceSerializer.serialize(documentReference));
+    }
+
+    private long getUploadMaxSize()
+    {
+        XWikiContext context = this.contextProvider.get();
+        return context.getWiki().getSpacePreferenceAsLong(UPLOAD_MAXSIZE_PARAMETER, UPLOAD_DEFAULT_MAXSIZE, context);
+    }
+
+    private int getCacheIdleTime()
+    {
+        return 3600;
+    }
+
+    private int getCacheMaxEntries()
+    {
+        return 10000;
+    }
+
+    private Cache<XWikiAttachment> getOrCreateCache(TemporaryAttachmentSession temporaryAttachmentSession,
+        DocumentReference documentReference) throws CacheException
+    {
+        Cache<XWikiAttachment> result;
+
+        if (!temporaryAttachmentSession.hasOpenEditionSession(documentReference)) {
+            String cacheName = getCacheName(temporaryAttachmentSession.getSessionId(), documentReference);
+            result = this.cacheManager.createNewCache(
+                new LRUCacheConfiguration(cacheName, getCacheMaxEntries(), getCacheIdleTime()));
```

Comment:
Why is this a cache with a maximum size? Do you need any caching behavior here that only files that are frequently accessed items are kept? This will cause absolutely weird bugs where users are randomly using temporary attachments if there are too many users uploading temporary attachments at the same time.

Reply:
Okay, maybe I misunderstood something here, this is a cache per user and document? If yes, I'm also not sure this is the right approach as I think having such a cache has quite some overhead (but I might be mistaken).

Reply:
> Okay, maybe I misunderstood something here, this is a cache per user and document?

yes

>  If yes, I'm also not sure this is the right approach as I think having such a cache has quite some overhead (but I might be mistaken).

Yes, this is what we've been discussing with @tmortagne on https://github.com/xwiki/xwiki-platform/commit/1681aa5e88c3d6764be948c3ca5970f47dec0541#r70594779 I'll change that approach.